granted credentials commands for securely storing IAM user credentials

meyerjrr opened the pull request:
Fixes #178 
- Renamed existing credstore code to tokenstore to better describe function
- Added credstore keychain which stores aws.credentials and retrieves when assuming a role
- if config field 'IAMCredStore' is set Granted will look for profiles/credentials in the keychain first then look in environment variables/config files
- Ability to add new credentials via the `granted credstore` command

chrnorm reviewed (dismissed):
Thanks for this @meyerjrr! I haven't looked at the code changes in depth but I have spun this up on my own machine to try it out. Some observations from me below.

## My testing setup

I created a test profile with plaintext credentials in `~/.aws/credentials`:
```
[test-iam-user]
aws_access_key_id=AKIAXXXXXXXX
aws_secret_access_key=XXXXXXXXXXXXX
region=us-east-1
```

## Testing

I was able to assume the IAM user just fine:
```
assume test-iam-user

[test-iam-user](us-east-1) session credentials ready
```

I then tried to add a user to the keychain:
```
❯ granted credstore add
Adding IAM credentials to `~/.aws/credentials is not yet supported by granted. UseSecureCredStorage: false⏎
```

I wasn't sure what to do here as the error message is unclear on how I can fix this - I had to look at the source code, and then I ran:

```
❯ granted credstore set
? Store credentials in your local keychain? Yes
IAM creds will be saved and read from your devices keychain⏎
```

I then ran:
```
❯ granted credstore add
IAM creds are being saved and read from your devices keychain
? Profile Name:  secure-iam-user

? Access Key Id:  AKIAY7GWILOWLLEZFA6I

? Secret Key:  ****************************************
Saved secure-iam-user to Credfile⏎
```

A note here: what is a "Credfile"? It sounds like a file being saved somewhere on my disk, but our implementation uses [keyring](https://pkg.go.dev/github.com/99designs/keyring).

After running this I then tried to assume the role, but was unable to:

```
❯ assume secure-iam-user
secure-iam-user does not match any profiles in your AWS config or credentials files
```

I then tried to assume the `test-iam-user` that I had in plaintext in my `~/.aws/credentials` file, and got a segfault:

```
❯ assume test-iam-user
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10099fa7c]

goroutine 21 [running]:
github.com/common-fate/granted/pkg/credstore.(*IAMUserProvider).Retrieve(0x0?, {0x0?, 0x0?})
	/Users/chrisnorman/code/granted/pkg/credstore/iam_creds.go:187 +0x4c
github.com/aws/aws-sdk-go-v2/aws.(*CredentialsCache).singleRetrieve(0x1400030a300, {0x100d84d38, 0x14000316390})
	/Users/chrisnorman/go/pkg/mod/github.com/aws/aws-sdk-go-v2@v1.14.0/aws/credential_cache.go:104 +0x78
github.com/aws/aws-sdk-go-v2/aws.(*CredentialsCache).Retrieve.func1()
	/Users/chrisnorman/go/pkg/mod/github.com/aws/aws-sdk-go-v2@v1.14.0/aws/credential_cache.go:89 +0x80
github.com/aws/aws-sdk-go-v2/internal/sync/singleflight.(*Group).doCall(0x1400030a330, 0x140001ad6e0, {0x0, 0x0}, 0x0?)
	/Users/chrisnorman/go/pkg/mod/github.com/aws/aws-sdk-go-v2@v1.14.0/internal/sync/singleflight/singleflight.go:97 +0x34
created by github.com/aws/aws-sdk-go-v2/internal/sync/singleflight.(*Group).DoChan
	/Users/chrisnorman/go/pkg/mod/github.com/aws/aws-sdk-go-v2@v1.14.0/internal/sync/singleflight/singleflight.go:90 +0x3c0
```

There is also no way to list or delete IAM users from the keychain?

Additionally, `granted credstore` shows confusing output, it seems to list the SSO instances:

```
❯ granted credstore
https://example.awsapps.com/start ()
https://other-instance.awsapps.com/start ()
```

Overall, I think the fact that we have a global config flag `UseSecureCredStorage` for using the keychain for IAM users is hurting the UX here. If I have plaintext credentials for `test-iam-user` in `~/.aws/credentials`, why does Granted prevent me from using it if `UseSecureCredStorage` is true?

Here's my proposed changes to improve UX here:

1. Remove the `UseSecureCredStorage` config flag. Users should be able to put IAM credentials in their keychain with `granted credstore add`, or in `~/.aws/credentials`.
2. Add `granted credstore import <PROFILE>` and `granted credstore export-plaintext <PROFILE>` commands. These commands should allow users to import and export credentials from `~/.aws/credentials`
3. When users use `assume` with a plaintext credential sourced from `~/.aws/credentials`, we print a message encouraging the user to import the credential into their keychain:
  ```
  Profile test-iam-user has plaintext credentials stored in ~/.aws/credentials. It's more secure to store these in your system's keychain. Move the credentials to your keychain with 'granted credstore import test-iam-user'.
  ```
4. Add `granted credstore list` and `granted credstore delete <PROFILE>` commands.
5. Make `granted credstore add` take a profile name as the argument, i.e. `granted credstore add <PROFILE>` - this is what users of `aws-vault` are used to.

meyerjrr commented:
Updated logic:
Removed config setting setting for where to store iam creds
Try keychain first, then check plaintext creds + give info/warning message about using keychain 
Add command for importing cred into keychain
Add profile name as command when adding new keychain profile

JoshuaWilkes commented on pkg/cfaws/assumer_aws_iam.go:
```diff
@@ -47,11 +69,18 @@ func (aia *AwsIamAssumer) AssumeTerminal(ctx context.Context, c *Profile, config
 		return aws.Credentials{}, err
 	}
+
+	//inform the user about securely securing iam users
```
```suggestion
	//inform the user about using the keychain to securely store IAM user credentials
```

JoshuaWilkes commented on pkg/cfaws/assumer_aws_iam.go:
```diff
@@ -47,11 +69,18 @@ func (aia *AwsIamAssumer) AssumeTerminal(ctx context.Context, c *Profile, config
+
+	//inform the user about securely securing iam users
+	fmt.Fprintf(color.Error, "Profile %s has plaintext credentials stored in ~/.aws/credentials.\n", c.Name)
+	fmt.Fprintf(color.Error, "It's more secure to store these in your system's keychain.\n")
```
```suggestion
	fmt.Fprintf(color.Error, "With Granted you can store these credentials more securely in your system's keychain.\n")
```

JoshuaWilkes reviewed (changes requested):
When a user shifts their credentials into the secure storage, It will break AWS commands that use a --profile flag.

These users will also need an entry in their config file with the profile name and a credential_process entry to use the granted credential process.


We should be able to automate all of this as part of the add/export commands

We need to test this for cases where there are credentials in the creds file and a matching profile in the config file which has additional config like regions etc.

You may also need to check whether the new credentials process code will work for IAM credentials.

JoshuaWilkes commented:
SecureStorage has been introduced to make the secure storage APIs more reusable.
```
SSOTokensSecureStorage
IAMCredentialsSecureStorage
```
The above provide a wrapper around secure storage with methods extending the functionality, this has been refactored in the codebase where it was used for SSO tokens.

> Note that I have renamed the servicename and folder used for tokens to be a composite of a prefix and suffix. The effect will be that existing granted users may be prompted once to allow granted access to the keychain, and they will need to sso login again to any profiles at most once.

I have proposed remaining the `granted token` command to `granted sso-tokens` because it makes it more clear now that we also have `granted credentials` 

I have chosen to refer to the secure storage as 'secure storage' rather than 'keychain' or 'keyring' because the implementation detail is dependent on each users machine. I propose that we also reword this anywhere it is used in the token commands.

JoshuaWilkes commented on pkg/granted/tokens.go:
```diff
@@ -9,14 +9,14 @@ import (
 var TokenCommand = cli.Command{
-	Name:        "token",
-	Usage:       "Manage aws access tokens",
+	Name:        "sso-tokens",
```
Note that I am proposing this command is renamed

JoshuaWilkes commented on pkg/securestorage/securestorage.go:
```diff
@@ -63,30 +110,30 @@ func openKeyring() (keyring.Keyring, error) {
-	credStorePath := path.Join(grantedFolder, "cred-store")
-
+	secureStoragePath := path.Join(grantedFolder, "secure-storage-"+s.StorageSuffix)
+	name := "granted-" + s.StorageSuffix
```
Note that renaming this will cause users to have to allow granted access to the keychain or setup a new secure password

JoshuaWilkes replied:
It was required to be able to have 2 seperate stores for credentials and sso tokens

JoshuaWilkes commented:
Docs PR here https://github.com/common-fate/docs/pull/121

chrnorm reviewed (changes requested):
Looking over the code the implementation looks good to me. Some feedback from my manual testing, mainly focused on UX:

- [x] let's customise the "usage" string in the new CLI commands to note that the profile is an argument that can be provided. I didn't know that it could be provided until I read the source code:

  ```diff
  ❯ dgranted credentials add --help
  NAME:
     granted credentials add - Add IAM credentials to secure storage
  
  USAGE:
  -   granted credentials add [command options] [arguments...]
  +   granted credentials add [<profile>]
  
  OPTIONS:
     --help, -h  show help (default: false)
  ```

- [x] I ran into a weird bug where I could get credentials exported as an empty profile:
  ```
  ❯ dgranted credentials import test-iam-user
  Saved test-iam-user to secure storage⏎
  
  ❯ dgranted credentials update
  ? Profile name:
  ? Access key id:
  ? Secret access key:
  Updated  in secure storage⏎
  
  ❯ dgranted credentials export-plaintext
  ? Profile name:
  Exported  in plaintext from secure storage to /Users/chrisnorman/.aws/credentials⏎
  
  ❯ cat ~/.aws/credentials
  []
  aws_access_key_id = AKIAXXXXXXXXXXX
  aws_secret_access_key = XXXXXXXXXXXXXXXX
  
  # these creds are for test-iam-user, somehow!
  ```

- [x] Overall I think we should enforce that the profile name is not an empty string (this may solve the problem above). I was able to run `granted credentials update` and `granted credentials export-plaintext` with an empty profile name
- [ ] `granted credentials export-plaintext` should print a message “the test-iam-user credentials have not been removed from secure storage. If you’d like to delete them, you can run `granted credentials remove test-iam-user`
- [ ] `granted credentials list`: if no credentials are found, we should print "No IAM user credentials stored in secure storage" to stderr

chrnorm commented:
Also, we have had some feedback from the users who requested this feature in #178:

> Awesome, this looks great! My only nitpick is that something like remove or delete is more intuitive (opposite of add) to me than clear.
> 
> aws-vault also provides a rotate command which is super handy, but more of a nice-to-have. It's like update, but it also takes care of generating the new keys at AWS.

> This is excellent. The only concern that I have is about this workflow -
> 
> I have an existing "hardcoded" AWS setup, with ~/.aws/config and ~/.aws/credentials
> I run granted credentials import
> I play around with granted and decide not to use it, or use it for a subset of profiles so I want to reset its storage after 10 minutes.
> I run granted credentials clear
> Concern: granted will clear ~/.aws/config and delete my original "hardcoded" setup
> If this is the case, it's could be quite a footgun because it will be difficult for me to recreate that hardcoded setup if I had that file for a while. I suppose the current solution will be to ensure I run granted credentials export-plaintext, so maybe the ...credentials clear should execute in dry-run mode by default and have a --force flag before it actually overrides the static config.

chrnorm commented on pkg/granted/credentials.go:
```diff
@@ -0,0 +1,441 @@
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Saved %s to secure storage", profileName)
```
A nitpick on the Printf statements: all of them in this file need a `\n` at the end so that they print a newline properly

JoshuaWilkes commented:
I have attempted to make this more fool proof by using select menus where an existing profile is required

I also improved the export and remove commands to work better together.
Now you can `export-plaintext` followed by `remove` and it will not delete the profile configuration, only the value in teh secure storage.

I have been able to safely run a lifecycle `add` -> `export-plaintext` -> `import` -> `export-plaintext` -> `remove`  
These sorts of combinations should all "work as expected"

chrnorm reviewed (changes requested):
Changes look good, I think we are very close here.

I imported some IAM user credentials with `dgranted credentials import example-creds` and then tried to assume the role, but got a region error:

```
❯ dassume example-creds
region not set on profile example-creds, could not load a default AWS_REGION. Either set a default region `aws configure set default.region us-west-2` or add a region to your profile
```

My `~/.aws/config` doesn't contain a `default` profile entry, it looks like this:

```
[profile example-creds]
credential_process = dgranted credential-process --profile=example-creds
region = us-west-2
```

When adding a default entry:
```
[profile example-creds]
credential_process = dgranted credential-process --profile=example-creds
region = us-west-2

[profile default]
region = ap-southeast-2
```

the role assumption works but it sets `AWS_REGION` to `ap-southeast-2`, rather than `us-west-2`. 

```
❯ dassume example-creds

[example-creds](ap-southeast-2) session credentials ready
```

I think what is happening is the credential-process command is not picking up the region from the profile, and trying to fall back to the default region.

So the last change to make here is to detect the region from the profile in the credential-process command.

chrnorm commented on pkg/cfaws/assumer_aws_iam.go:
```diff
@@ -41,17 +50,25 @@ func (aia *AwsIamAssumer) AssumeTerminal(ctx context.Context, c *Profile, config
+
+	//inform the user about using the secure storage to securely store IAM user credentials
+
+	fmt.Fprintf(color.Error, "Profile %s has plaintext credentials stored in the AWS credentials file.\n", c.Name)
+	fmt.Fprintf(color.Error, "With Granted you can store these credentials in secure storage.\n")
+	fmt.Fprintf(color.Error, "To move the credentials to secure storage, run `granted credentials import %s`.\n", c.Name)
```
These prompts don't appear for me in my testing, @JoshuaWilkes have you seen them in your local dev environments?

JoshuaWilkes replied:
Fixed in latest changes
